src/mscorlib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -388,13 +388,15 @@ private bool TryInsert(TKey key, TValue value, InsertionBehavior behavior)
             int targetBucket = hashCode % buckets.Length;
             int collisionCount = 0;
 
-            for (int i = buckets[targetBucket]; i >= 0; i = entries[i].next)
+            int i = buckets[targetBucket];
+            while (i >= 0)
             {
-                if (entries[i].hashCode == hashCode && comparer.Equals(entries[i].key, key))
+                ref Entry candidateEntry = ref entries[i];
+                if (candidateEntry.hashCode == hashCode && comparer.Equals(candidateEntry.key, key))
                 {
                     if (behavior == InsertionBehavior.OverwriteExisting)
```

Comment:
@benaadams Here you can use the marker interface on the caller to evict the whole comparison in the loop. 

On the call site use ```TryInsert<OverwriteExisting>(...)``` and ```TryInsert<ThrowOnExisting>(...)``` and generate 2 separate codepath for each.

Reply:
I did that in https://github.com/dotnet/coreclr/pull/14030; however its a bit trickier as dictionary is heavily crossgened and the whole stack is AoT R2R (rather than Jit) for the runtime store which cuts out some optimizations and introduces speed bumps for the class path.

I think the better approach is actually split the method and not use the flag

Reply:
In that case, probably, yes.